### PR TITLE
Migrate decimal and experimental to Preferences

### PR DESCRIPTION
**Pull request by zbraniecki**
Migrated experimental formatters to Preferences

### CODE REVIEW

File: components/experimental/src/compactdecimal/formatter.rs [hunk 1]
```diff
@@ -15,11 +15,18 @@ use alloc::borrow::Cow;
...
+use icu_locale_core::preferences::define_preferences;
 use icu_plurals::PluralRules;
 use icu_provider::DataError;
 use icu_provider::{marker::ErasedMarker, prelude::*};
 use zerovec::maps::ZeroMap2dCursor;
 
+define_preferences!(
+    /// The preferences for compact decimal formatting.
+    CompactDecimalFormatterPreferences,
+    {}
```
Comment (sffc) [inline components/experimental/src/compactdecimal/formatter.rs:27]: I think this should retain the numbering system

File: components/experimental/src/compactdecimal/formatter.rs [hunk 2]
```diff
@@ -78,26 +85,28 @@ impl CompactDecimalFormatter {
...
         options: CompactDecimalFormatterOptions,
     ) -> Result<Self, DataError> {
-        let temp_loc = locale.clone().into_locale();
+        let locale = DataLocale::from_preferences_locale::<ShortCompactDecimalFormatDataV1Marker>(
+            prefs.locale_prefs,
+        );
         Ok(Self {
             fixed_decimal_formatter: FixedDecimalFormatter::try_new(
-                locale,
+                prefs.locale_prefs.into(),
```
Comment (sffc) [inline components/experimental/src/compactdecimal/formatter.rs:102]: And then down here, forward the numbering system into the FixedDecimalFormatter

File: components/locale_core/src/preferences/mod.rs [hunk 3]
```diff
@@ -529,6 +529,18 @@ macro_rules! __define_preferences {
             }
         }
 
+        impl From<$crate::preferences::LocalePreferences> for $name {
+            fn from(locale_prefs: $crate::preferences::LocalePreferences) -> Self {
```
Comment (sffc) [inline components/locale_core/src/preferences/mod.rs:533]: Issue: I don't know if I am onboard with this impl. It makes it easy to forget to forward the extensions from one preference bag to another preference bag. This is a source of bugs in ICU.

File: components/experimental/src/dimension/currency/formatter.rs [hunk 4]
```diff
@@ -15,6 +16,12 @@ use super::CurrencyCode;
 
 extern crate alloc;
 
+define_preferences!(
+    /// The preferences for currency formatting.
+    CurrencyFormatterPreferences,
+    {}
```
Comment (sffc) [inline components/experimental/src/dimension/currency/formatter.rs:22]: Issue: This one should definitely take the numbering system.

File: components/experimental/src/dimension/percent/formatter.rs [hunk 5]
```diff
@@ -17,6 +18,12 @@ use super::options::PercentFormatterOptions;
 
 extern crate alloc;
 
+define_preferences!(
+    /// The preferences for percent formatting.
+    PercentFormatterPreferences,
+    {}
+);
+
```
Comment (sffc) [inline components/experimental/src/dimension/percent/formatter.rs:26]: Issue: Numbering system

**Review by sffc — CHANGES_REQUESTED** [review]
Yeah, so all the things that forward to FixedDecimalFormatter should retain the numbering system from the input locale. `impl From<LocalePreferences> for *Preferences` is a footgun because it drops those extensions.

File: components/locale_core/src/preferences/mod.rs (hunk 3 above)
Comment (zbraniecki) [reply components/locale_core/src/preferences/mod.rs]: that makes sense. I replaced it with a macro to generate conversions.

File: components/experimental/src/compactdecimal/formatter.rs [hunk 6]
```diff
@@ -14,12 +14,30 @@ use crate::compactdecimal::{
...
+        numbering_system: NumberingSystem
+    }
+);
+
+prefs_convert!(
+    CompactDecimalFormatterPreferences,
+    FixedDecimalFormatterPreferences,
+    { numbering_system }
+);
+prefs_convert!(CompactDecimalFormatterPreferences, PluralRulesPreferences);
```
Comment (sffc) [inline components/experimental/src/compactdecimal/formatter.rs:39]: Praise: This `prefs_convert` thing seems like a good solution

**Review by sffc — APPROVED** [review]
Nice work!

**Pull request merged**